ts/deployment/deployDeployer.ts

```diff
@@ -1,80 +1,42 @@
 import { Signer } from "ethers";
-import { BigNumber } from "ethers";
 import assert from "assert";
-import { keylessDeploy, calculateKeylessDeployment } from "./keylessDeployment";
-import { Provider } from "@ethersproject/providers";
+import { KeylessDeployer } from "./keylessDeployment";
 import {
     deployerBytecode,
     DEPLOYER_ADDRESS,
     KEYLESS_DEPLOYMENT
 } from "./static";
 import { logAddress, logDeployment } from "../../scripts/logger";
 
-export async function calculateDeployerAddress(
-    provider?: Provider
-): Promise<{ deployerAddress: string; keylessAccount: string }> {
-    let result = await calculateKeylessDeployment(
-        provider,
-        deployerBytecode(),
-        KEYLESS_DEPLOYMENT.GAS_PRICE,
-        KEYLESS_DEPLOYMENT.GAS_LIMIT,
-        false
-    );
-    return {
-        deployerAddress: result.contractAddress,
-        keylessAccount: result.keylessAccount
-    };
-}
-
-export async function calculateGasLimit(
-    provider: Provider
-): Promise<BigNumber> {
-    let result = await calculateKeylessDeployment(
-        provider,
-        deployerBytecode(),
-        KEYLESS_DEPLOYMENT.GAS_PRICE,
-        KEYLESS_DEPLOYMENT.GAS_LIMIT,
-        false
-    );
-    return result.estimatedGasCost;
-}
-
 export async function deployDeployer(
     signer: Signer,
     verbose: boolean
 ): Promise<boolean> {
     assert(signer.provider);
     const provider = signer.provider;
-    const result = await calculateKeylessDeployment(
-        provider,
+    // Need exact GAS_PRICE and GAS_LIMIT to derive correct DEPLOYER_ADDRESS
+    const deployer = new KeylessDeployer(
         deployerBytecode(),
         KEYLESS_DEPLOYMENT.GAS_PRICE,
-        KEYLESS_DEPLOYMENT.GAS_LIMIT,
-        verbose
-    );
+        KEYLESS_DEPLOYMENT.GAS_LIMIT
+    ).connect(provider);
 
-    if (result.alreadyDeployed) {
+    if (await deployer.alreadyDeployed()) {
         logAddress(
             verbose,
             "Deployer is ALREADY deployed",
-            result.contractAddress
+            deployer.contractAddress
         );
         return true;
     }
-    assert(KEYLESS_DEPLOYMENT.GAS_LIMIT.gte(result.estimatedGasCost));
-    assert(DEPLOYER_ADDRESS == result.contractAddress);
-    const _result = await keylessDeploy(
-        signer,
-        deployerBytecode(),
-        KEYLESS_DEPLOYMENT.GAS_PRICE,
-        KEYLESS_DEPLOYMENT.GAS_LIMIT,
-        verbose
-    );
+    assert(DEPLOYER_ADDRESS == deployer.contractAddress);
```

Comment:
We here check only DEPLOYER_ADDRESS because that's what matters.
We can't do `assert(KEYLESS_DEPLOYMENT.GAS_LIMIT.gte(await deployer.estimateGas()));` here because at the stage, we haven't funded the keylessAccount yet, so the estimateGas will fail for the insufficient fund. Hardhat seems to be fine with this, but Geth is strict 😢